lib/HLSL/DxilScalarizeVectorLoadStores.cpp

```diff
@@ -0,0 +1,239 @@
+//                                                                           //
+// DxilScalarizeVectorLoadStores.cpp                                         //
+// Copyright (C) Microsoft Corporation. All rights reserved.                 //
+// This file is distributed under the University of Illinois Open Source     //
+// License. See LICENSE.TXT for details.                                     //
+//                                                                           //
+// Lowers native vector load stores to potentially multiple scalar calls.    //
+//                                                                           //
+///////////////////////////////////////////////////////////////////////////////
+
+#include "dxc/DXIL/DxilInstructions.h"
+#include "dxc/DXIL/DxilModule.h"
+#include "dxc/HLSL/DxilGenerationPass.h"
+
+#include "llvm/ADT/StringRef.h"
+#include "llvm/IR/Function.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Instructions.h"
+#include "llvm/IR/Module.h"
+#include "llvm/IR/PassManager.h"
+#include "llvm/Pass.h"
+
+using namespace llvm;
+using namespace hlsl;
+
+class DxilScalarizeVectorLoadStores : public ModulePass {
+private:
+  DxilModule *m_DM;
+
+  void scalarizeVectorLoad(hlsl::OP *HlslOP, const DataLayout &DL,
+                           CallInst *CI);
+  void scalarizeVectorStore(hlsl::OP *HlslOP, const DataLayout &DL,
+                            CallInst *CI);
+
+public:
+  static char ID; // Pass identification, replacement for typeid
+  explicit DxilScalarizeVectorLoadStores() : ModulePass(ID) {}
+
+  StringRef getPassName() const override {
+    return "DXIL scalarize vector load/stores";
+  }
+
+  bool runOnModule(Module &M) override {
+    DxilModule &DM = M.GetOrCreateDxilModule();
+    m_DM = &DM;
+
+    // Shader Model 6.9 allows native vectors and doesn't need this pass.
+    if (DM.GetShaderModel()->IsSM69Plus())
+      return false;
+
+    bool Changed = false;
+
+    hlsl::OP *HlslOP = DM.GetOP();
+    for (auto F = M.functions().begin(), E = M.functions().end(); F != E;) {
+      Function *Func = &*(F++);
+      DXIL::OpCodeClass OpClass;
+      if (HlslOP->GetOpCodeClass(Func, OpClass)) {
+        if (OpClass == DXIL::OpCodeClass::RawBufferVectorLoad) {
+          for (auto U = Func->user_begin(), UE = Func->user_end(); U != UE;) {
+            CallInst *CI = cast<CallInst>(*(U++));
+            scalarizeVectorLoad(HlslOP, M.getDataLayout(), CI);
+            Changed = true;
+          }
+          Func->eraseFromParent();
+        } else if (OpClass == DXIL::OpCodeClass::RawBufferVectorStore) {
+          for (auto U = Func->user_begin(), UE = Func->user_end(); U != UE;) {
+            CallInst *CI = cast<CallInst>(*(U++));
+            scalarizeVectorStore(HlslOP, M.getDataLayout(), CI);
+            Changed = true;
+          }
+          Func->eraseFromParent();
+        }
+      }
+    }
+    return Changed;
+  }
+};
+
+static unsigned GetRawBufferMask(unsigned NumComponents) {
```

Comment:
Of course, this could have been just `(1 << NumComponents) - 1` (with assert for `NumComponents <= 4`).  But oh well.